app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt

```diff
@@ -83,48 +122,41 @@ class DefaultTasksRepository(
         if (forceUpdate) {
             updateTaskFromRemoteDataSource(taskId)
         }
-        return tasksLocalDataSource.getTask(taskId)
-    }
-
-    override suspend fun saveTask(task: Task) {
-        coroutineScope {
-            launch { tasksRemoteDataSource.saveTask(task) }
-            launch { tasksLocalDataSource.saveTask(task) }
-        }
+        return tasksDao.getTaskById(taskId)?.asExternalModel()
     }
 
     override suspend fun completeTask(taskId: String) {
         coroutineScope {
-            launch { tasksRemoteDataSource.completeTask(taskId) }
-            launch { tasksLocalDataSource.completeTask(taskId) }
+            launch { tasksNetworkDataSource.completeTask(taskId) }
```

Comment:
The codependency dependency of the coroutines launched here are a bit hard to reconcile. If any coroutine fails, while the other is running, the other one is cancelled. However either can succeed first, and the other can fail next.

I think it'd be a bit clearer if both must succeed, or if one fails, it doesn't force the other to be cancelled if it is still running.